diff/diff.go

```diff
@@ -0,0 +1,534 @@
+			return false, errors.Trace(err)
+		}
+		defer rows2.Close()
+
+		eq, err := df.compareRows(rows1, rows2, orderKeyCols, table)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		if !eq {
+			equal = false
+		}
+	}
+
+	return equal, nil
+}
+
+func (df *Diff) compareRows(rows1, rows2 *sql.Rows, orderKeyCols []*model.ColumnInfo, table *TableCheckCfg) (bool, error) {
+	equal := true
+	rowsData1 := make([]map[string][]byte, 0, 100)
+	rowsData2 := make([]map[string][]byte, 0, 100)
+
+	for rows1.Next() {
+		data1, err := pkgdb.ScanRow(rows1)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		rowsData1 = append(rowsData1, data1)
+	}
+	for rows2.Next() {
+		data2, err := pkgdb.ScanRow(rows2)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		rowsData2 = append(rowsData2, data2)
+	}
+
+	var index1, index2 int
+	for {
+		if index1 == len(rowsData1) {
+			// all the rowsData2's data should be deleted
+			for ; index2 < len(rowsData2); index2++ {
+				sql := generateDML("delete", rowsData2[index2], orderKeyCols, table.Info, table.Schema)
+				log.Infof("[delete] sql: %v", sql)
+				df.wg.Add(1)
+				df.sqlCh <- sql
+				equal = false
+			}
+			break
+		}
+		if index2 == len(rowsData2) {
+			// rosData2 lack some data, should insert them
+			for ; index1 < len(rowsData1); index1++ {
+				sql := generateDML("replace", rowsData1[index1], orderKeyCols, table.Info, table.Schema)
+				log.Infof("[insert] sql: %v", sql)
+				df.wg.Add(1)
+				df.sqlCh <- sql
+				equal = false
+			}
+			break
+		}
+		eq, cmp, err := compareData(rowsData1[index1], rowsData2[index2], orderKeyCols)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		if eq {
+			index1++
+			index2++
+			continue
+		}
+		equal = false
+		switch cmp {
+		case 1:
+			// delete
+			sql := generateDML("delete", rowsData2[index2], orderKeyCols, table.Info, table.Schema)
+			log.Infof("[delete] sql: %v", sql)
+			df.wg.Add(1)
+			df.sqlCh <- sql
+			index2++
+		case -1:
+			// insert
+			sql := generateDML("replace", rowsData1[index1], orderKeyCols, table.Info, table.Schema)
+			log.Infof("[insert] sql: %v", sql)
+			df.wg.Add(1)
+			df.sqlCh <- sql
+			index1++
+		case 0:
+			// update
+			deleteSQL := generateDML("delete", rowsData2[index2], orderKeyCols, table.Info, table.Schema)
+			replaceSQL := generateDML("replace", rowsData1[index1], orderKeyCols, table.Info, table.Schema)
+			log.Infof("[update] delete: %s,\n replace: %s", deleteSQL, replaceSQL)
+			df.wg.Add(2)
+			df.sqlCh <- deleteSQL
+			df.sqlCh <- replaceSQL
+			index1++
+			index2++
+		}
+	}
+
+	return equal, nil
+}
+
+// WriteSqls write sqls to file
+func (df *Diff) WriteSqls() {
+	defer df.wg.Done()
+	for {
+		select {
+		case dml, ok := <-df.sqlCh:
+			if !ok || dml == "end" {
+				return
+			}
+
+			_, err := df.fixSQLFile.WriteString(fmt.Sprintf("%s\n", dml))
+			if err != nil {
+				log.Errorf("write sql: %s failed, error: %v", dml, err)
+			}
+			df.wg.Done()
+		default:
```

Comment:
use `default` to run the `for` loop without blocking?

Reply:
you suggest remove it?

Reply:
yes